ops/model.py

```diff
@@ -476,6 +480,45 @@ def location(self):
         return self._location
 
 
+class Function:
+    def __init__(self, backend):
+        self.name = backend.function_name
+        self.parameters = FunctionParameters(backend)
+        self.results = FunctionResults(backend)
+        self._backend = backend
+
+    def fail(self, message=''):
+        self._backend.function_fail(message)
+
+
+class FunctionParameters(LazyMapping):
+
+    def __init__(self, backend):
+        self._backend = backend
+
+    def _load(self):
+        return self._backend.function_get()
+
+
+class FunctionResults(LazyMapping, MutableMapping):
```

Comment:
As we discussed in our call today, this doesn't feel right for functions. The context of the function is only valid while the event is being handled. Once the event is over, juju will consider the function as terminated and these parameters and results are meaningless.

This would probably be better modeled as a richer event type.